uwblueprint/starter-code-v2
# Integrate Google OAuth into Python backend

## Notion ticket link

Closes #103 


## Implementation description
* Added some frontend (REST API calls) to generate the OAuth id token for testing
* Backend uses the OAuth id token to get the associated Google user using new firebase rest client method
* Once we get the Google user info, generate access/refresh tokens
* If the user already has an account with that email, logs them in
* Otherwise, creates a new account with that email
* A lot of the code is based off the Planet Read PR (linked in issue), but simplified for more general use case 


## Steps to test
1. Add FIREBASE_REQUEST_URI=http://localhost to backend .env. 
Also, go into GCP console -> APIs & Services -> Credentials and copy the client id for Web Client 2. Add REACT_APP_OAUTH_CLIENT_ID=\<your client id\> to your frontend .env. 
Also, make sure MG_DATABASE_URL in your backend .env is pointing to python-test database.
<img width="1210" alt="image" src="https://user-images.githubusercontent.com/37782734/130537538-99be214f-e254-4268-950a-b029b5a1f303.png"> . 
2. Go to GCP console -> APIs & Services -> OAuth consent screen. Add your personal email to Test users. This allows you to use your personal email for testing sign in (otherwise you can only sign in with emails from blueprint org).
<img width="490" alt="image" src="https://user-images.githubusercontent.com/37782734/130537553-9f2e6308-f282-448c-81d1-c4be315aa719.png">

3. For the frontend files, uncomment the REST sections added in this PR and comment the corresponding GraphQL sections. Uncomment any other REST sections needed in order to get the frontend to compile (eg. importing baseApiClient in AuthAPIClient.ts).
4. When the frontend is compiled successfully, go to the login screen. Click on Login with Google and choose your uwblueprint email. Verify that you are logged in.
<img width="365" alt="image" src="https://user-images.githubusercontent.com/37782734/130537574-26ffd89c-92df-484f-b764-faff23449b43.png">


<img width="912" alt="image" src="https://user-images.githubusercontent.com/37782734/130538876-85fccb87-8ea4-46e4-bd74-c7d6dcd63660.png">



<img width="956" alt="image" src="https://user-images.githubusercontent.com/37782734/130537952-4eb8ba1d-62c2-4a81-afb8-713313a0765e.png">

5. Logout, and go back to the login screen. Make sure that your personal email is deleted from Firebase (Firebase -> Build -> Authentication) and on MongoDB atlas/postgres. This time, try logging in with your personal email. Verify that you are logged in, a User is created for you in MongoDB atlas/postgres based on the Google account info, and a user is created in Firebase.
<img width="428" alt="image" src="https://user-images.githubusercontent.com/37782734/130538017-924a5624-a4da-4915-bdea-e5fabec6da94.png">



## What should reviewers focus on?
* Did I miss any cases?


## Checklist
